platform/ui/src/contextProviders/ViewportGridProvider.tsx

```diff
@@ -56,32 +67,32 @@ const reuseViewport = (idSet, viewport, stateViewports) => {
     ) {
       continue;
     }
+    if (idSet[oldId]) continue;
     idSet[oldId] = true;
-    // TODO re-use viewports once the flickering/wrong size redraw is fixed
-    // return {
-    //   ...oldViewport,
-    //   ...viewport,
-    //   viewportOptions: {
-    //     ...oldViewport.viewportOptions,
-
-    //     viewportId: oldViewport.viewportId,
-    //   },
-    // };
-  }
-  // Find a viewport instance number different from earlier viewports having
-  // the same presentationIds as this one would - will be less than 10k
-  // viewports hopefully :-)
-  for (let i = 0; i < 10000; i++) {
-    const viewportId = 'viewport-' + i;
-    if (idSet[viewportId] || oldIds[viewportId]) continue;
-    idSet[viewportId] = true;
-    return {
-      ...viewport,
-      viewportId,
-      viewportOptions: { ...viewport.viewportOptions, viewportId },
-    };
+    if (isReuseableViewport(oldViewport, viewport)) {
+      return {
+        ...oldViewport,
+        ...viewport,
+        id: oldViewport.viewportId,
+        viewportOptions: {
+          ...oldViewport.viewportOptions,
+
+          viewportId: oldViewport.viewportId,
+        },
+      };
+    }
   }
-  throw new Error('No ID found');
+  // Find a viewport instance number different from earlier viewports
+  const viewportId = 'viewport-' + viewportCounter;
```

Comment:
does it makes sense instead of letting viewportId just goes to infinit (100000), we create it based on displaySetInstanceUIDs?

Reply:
That causes all sorts of issue - you then need to figure out a unique addendum to the viewport to make it unique again because you can have two viewports shown at once.  Also, during race conditions where there are lots of updates to viewports you get some ugly errors.